python/mlx/nn/layers/linear.py

```diff
@@ -26,20 +41,92 @@ class Linear(Module):
+        self.output_dims = output_dims
+        self.weight = mx.zeros((output_dims, input_dims))
+        if bias:
+            self.bias = mx.zeros((output_dims,))
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        scale = math.sqrt(1.0 / self.input_dims)
         self.weight = mx.random.uniform(
             low=-scale,
             high=scale,
-            shape=(output_dims, input_dims),
+            shape=(self.output_dims, self.input_dims),
         )
+        if "bias" in self:
+            self.bias = mx.random.uniform(
+                low=-scale,
+                high=scale,
+                shape=(self.output_dims,),
+            )
+
+    def _extra_repr(self):
+        return f"input_dims={self.input_dims}, output_dims={self.output_dims}, bias={'bias' in self}"
+
+    def __call__(self, input):
+        output = input @ self.weight.T
+        if "bias" in self:
+            output = output + self.bias
+        return output
+
+
+class Bilinear(Module):
+    r"""Applies a bilinear transformation to the input.
+
+    Concretely:
+
+    .. math::
+
+        y = input1^\top W input2 + b
+
+    where :math:`W` has shape ``[output_dims, input1_dims, input2_dims]``.
+
+    Args:
+        input1_dims (int): The dimensionality of the input1 features
+        input2_dims (int): The dimensionality of the input2 features
+        output_dims (int): The dimensionality of the output features
+        bias (bool, optional): If set to ``False`` then the layer will
+          not use a bias. Default ``True``.
+    """
+
+    def __init__(
+        self, input1_dims: int, input2_dims: int, output_dims: int, bias: bool = True
+    ):
+        super().__init__()
+        self.input1_dims = input1_dims
+        self.input2_dims = input2_dims
+        self.output_dims = output_dims
+        self.weight = mx.zeros((output_dims, input1_dims, input2_dims))
         if bias:
             self.bias = mx.zeros((output_dims,))
 
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        scale = math.sqrt(1.0 / self.input1_dims)
+        self.weight = mx.random.uniform(
+            low=-scale,
+            high=scale,
+            shape=(self.output_dims, self.input1_dims, self.input2_dims),
+        )
+        if "bias" in self:
+            self.bias = mx.random.uniform(
+                low=-scale,
+                high=scale,
+                shape=(self.output_dims,),
+            )
+
     def _extra_repr(self):
-        return f"input_dims={self.weight.shape[1]}, output_dims={self.weight.shape[0]}, bias={'bias' in self}"
+        return (
+            f"input1_dims={self.input1_dims}, input2_dims={self.input2_dims}, output_dims={self.output_dims}, "
+            f"bias={'bias' in self}"
+        )
 
-    def __call__(self, x):
-        x = x @ self.weight.T
+    def __call__(self, input1, input2):
+        output = (input1 @ self.weight * input2.reshape(1, *input2.shape)).sum(-1).T
```

Comment:
This should be done with two matrix multiplications not a multiplication + sum.

Reply:
Hi, I think the math is ok [(reference)](https://www.quora.com/What-is-a-bilinear-tensor-layer-in-contrast-to-a-standard-linear-neural-network-layer-or-how-can-I-imagine-it) and the implementation aligns with earlier versions of torch (https://github.com/pytorch/pytorch/pull/1146, https://github.com/pytorch/pytorch/pull/1104).

I've implemented several versions and compared them to `torch.nn.functional.bilinear`. To avoid using a for loop, I created a custom implementation that could be replaced with `einsum` in the future.

Let me know your thoughts and if I missed something.

```python
import torch
import torch.nn.functional as F

# Set deterministic algorithms for reproducibility
torch.use_deterministic_algorithms(True)
torch.manual_seed(0)
torch.backends.cudnn.deterministic = True

# Generate random tensors
x1 = torch.randn((10, 20))
x2 = torch.randn((10, 40))
w = torch.randn((60, 20, 40))

# Original implementation using F.bilinear
res_original = F.bilinear(x1, x2, w)

# Naive implementation
res_naive= torch.cat([(x1.mm(w[i]).mul(x2)).sum(1, keepdims=True) for i in range(w.shape[0])], 1)

# Einsum implementation
res_einsum = torch.einsum('bi,kij,bj->bk', x1, w, x2)

# Custom implementation
res_custom= (((torch.matmul(x1, w) * x2.unsqueeze(0))).sum(-1)).T

# Print mean absolute differences between implementations
print('Mean Absolute Difference - Naive:', abs(res_original - res_naive).mean())
print('Mean Absolute Difference - Einsum:', abs(res_original - res_einsum).mean())
print('Mean Absolute Difference - Custom:', abs(res_original - res_custom).mean())
```

Output:
```python
Mean Absolute Difference - Naive: tensor(2.7407e-06)
Mean Absolute Difference - Einsum: tensor(6.3483e-06)
Mean Absolute Difference - Custom: tensor(2.7407e-06)
```

Reply:
Right I did not mean to imply it was in pro. Just that it will be a lot faster to use two matrix multiplications instead of a multiply and a reduce. Can you switch it?

Reply:
If I understand correctly, you're asking for a bilinear transformation using two matrix multiplications. However, this leads to extra calculations and a longer runtime due to the bilinear operation being only the diagonal of `x1 @ w @ x2.T`. I've removed unnecessary reshaping in the latest commit. Until we have the einsum operation, I believe the method I've used is the most efficient. Let me know if I missed anything.

Reply:
Assuming the inputs have the dimensions expanded appropriately:

- `x1` has shape `[B, 1, 1, D1]`
- `W` has shape `[1, D, D1, D2]`
- `x2` has shape `[B, 1, D2]`

You can do something like:

```
y = mx.squeeze(x1 @ w, -2) # shape B x D x D2
z = x2 @ y.swapaxes(-1, -2) # shape B x 1 x D
return mx.squeeze(z, -2)
```

You might play around with it and find a simpler implementation.

Reply:
Thanks for writing the implementation; I haven't thought of it. I made a runtime comparison using `timeit` with the conditions `B=100, D=60, D1=20, D2=40`, and my implementation was >20% faster. Let me know if you still want me to change this.

Reply:
> and my implementation was >20% faster. Let me know if you still want me to change this.

Can you post the benchmark? That doesn't make sense that the broadcast + reduce is faster. 

It's possible for really small `D*` it is comparable but for large `D*` using a matmul should be significantly faster and also use way less memory.

Reply:
Hi, Thanks for catching that! I took another test, and you're right. I've made the changes you asked for in [this commit](https://github.com/ml-explore/mlx/pull/315/commits/8e22ab1e5a3ebb8e79ed4273dac9273b7e223bb6). Let me know if everything looks good now.